sql/binlog.cc

```diff
@@ -5013,14 +5013,20 @@ bool MYSQL_BIN_LOG::check_write_error(THD *thd) {
 
   if (!thd->is_error()) DBUG_RETURN(checked);
 
-  switch (thd->get_stmt_da()->mysql_errno()) {
-    case ER_TRANS_CACHE_FULL:
-    case ER_STMT_CACHE_FULL:
-    case ER_ERROR_ON_WRITE:
-    case ER_BINLOG_LOGGING_IMPOSSIBLE:
-      checked = true;
-      break;
+  // Check all conditions for one that matches the expected error
+  const Sql_condition *err;
+  auto it = thd->get_stmt_da()->sql_conditions();
+  while ((err = it++) != nullptr && !checked) {
```

Comment:
Percona's fix also checks `thd->get_stmt_da()->mysql_errno()` in addition to iterating through `thd->get_stmt_da()->sql_conditions()`. Could you confirm that checking only `thd->get_stmt_da()->sql_conditions()` is enough.

Reply:
I looks that checking only `thd->get_stmt_da()->sql_conditions()` should be enough:
1. `m_mysql_errno` is changed only in `set_error_status()` and `reset_diagnostics_area()'.
In both cases `m_status` is also changed (to `DA_ERROR` and `DA_EMPTY` respectively)
2. `check_write_error()` checks if 
`if (!thd->is_error()) DBUG_RETURN(checked);`
and we have
```
bool is_error() const { return m_status == DA_ERROR; }
```